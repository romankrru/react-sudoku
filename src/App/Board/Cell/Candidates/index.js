import React, { memo } from "react";
import _ from "lodash";
import cn from "classnames";
import T from "prop-types";

import styles from "./index.module.css";

const Candidates = (props) => (
	<div className={styles.candidates}>
		{_.range(1, 10).map((n) => {
			return (
				<div
					onMouseDown={() => {
						if (props.isMobile || !props.isFocus) return;
						props.toggleCandidate(props.rowIdx, props.colIdx, n);
					}}
					key={n}
					className={cn(styles.candidateCell, {
						[styles.candidateCellActive]: props.candidates.includes(n),
						[styles.candidateCellHidden]:
							!props.candidates.includes(n) && !props.isFocus,
					})}
				>
					<span>{n}</span>
				</div>
			);
		})}
	</div>
);

Candidates.propTypes = {
	isFocus: T.bool.isRequired,
	toggleCandidate: T.func.isRequired,
	rowIdx: T.number.isRequired,
	colIdx: T.number.isRequired,
	isMobile: T.bool.isRequired,
	candidates: T.arrayOf(T.number).isRequired,
};

export default memo(Candidates);
