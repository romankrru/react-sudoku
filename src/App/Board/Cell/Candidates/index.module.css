.candidates {
	display: grid;
	grid-template-columns: repeat(3, 1fr);
	grid-template-rows: repeat(3, 1fr);

	width: 100%;
	height: 100%;

	font-size: 11px;
	color: transparent;
}

.candidates:hover {
	color: #c4c4c4;
}

.candidateCell {
	display: flex;
	width: 100%;
	height: 100%;
	justify-content: center;
	align-items: center;

	transition: all 0.3s ease;
}

.candidateCell:hover {
	color: #7eaefc;
}

.candidateCellActive {
	color: #2979fa;
}

.candidateCellActive:hover {
	color: #2979fa;
}

.candidateCellHidden {
	color: transparent;
}

.candidateCellHidden:hover {
	color: transparent;
}
