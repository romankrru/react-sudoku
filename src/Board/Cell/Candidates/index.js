import React from "react";
import cn from "classnames";
import styles from "./index.module.css";

const range = Array.from({ length: 9 }, (_, i) => i + 1);

const Candidates = (props) => {
	return (
		<div className={styles.candidates}>
			{range.map((n) => {
				return (
					<div
						{...(props.isFocus && {
							onClick: () =>
								props.toggleCandidate(props.rowIdx, props.colIdx, n),
						})}
						key={n}
						className={cn(styles.candidateCell, {
							[styles.candidateCellActive]: props.candidates.includes(n),
							[styles.candidateCellHidden]:
								!props.candidates.includes(n) && !props.isFocus,
						})}
					>
						{n}
					</div>
				);
			})}
		</div>
	);
};

export default Candidates;
