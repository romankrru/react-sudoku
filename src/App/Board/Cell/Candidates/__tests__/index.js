import React from "react";
import { render, fireEvent } from "@testing-library/react";

import Candidates from "../index";

const renderCandidates = () => {
	const defaultProps = {
		isFocus: false,
		toggleCandidate: jest.fn,
		rowIdx: 0,
		colIdx: 0,
		isMobile: false,
		candidates: [],
	};

	const view = render(<Candidates {...defaultProps} />);

	const rerender = (newProps) =>
		view.rerender(<Candidates {...{ ...defaultProps, ...newProps }} />);

	return { view, rerender };
};

describe("<Candidates />", () => {
	it("should show correct candidates", () => {
		const { view, rerender } = renderCandidates();

		// empty candidates
		expect(
			view.container.querySelectorAll(".candidateCellHidden")
		).toHaveLength(9);

		expect(
			view.container.querySelectorAll(".candidateCellActive")
		).toHaveLength(0);

		// [1, 3, 6]
		rerender({ candidates: [1, 3, 6] });

		expect(
			view.container.querySelectorAll(".candidateCellHidden")
		).toHaveLength(6);

		expect(
			view.container.querySelectorAll(".candidateCellActive")
		).toHaveLength(3);

		expect(view.container).toMatchSnapshot();

		// [7]
		rerender({ candidates: [7] });

		expect(
			view.container.querySelectorAll(".candidateCellHidden")
		).toHaveLength(8);

		expect(
			view.container.querySelectorAll(".candidateCellActive")
		).toHaveLength(1);

		expect(
			view.container.querySelector(".candidateCellActive").textContent
		).toBe("7");

		// all candidates
		rerender({ candidates: [1, 2, 3, 4, 5, 6, 7, 8, 9] });

		expect(
			view.container.querySelectorAll(".candidateCellHidden")
		).toHaveLength(0);

		expect(
			view.container.querySelectorAll(".candidateCellActive")
		).toHaveLength(9);
	});

	it("should handle mousedown event correctly", () => {
		const { view, rerender } = renderCandidates();

		const handlerMock = jest.fn(() => {});

		rerender({
			isFocus: true,
			isMobile: false,
			toggleCandidate: handlerMock,
		});

		// find first candidate cell
		fireEvent.mouseDown(view.container.querySelector(".candidateCell"));
		expect(handlerMock).toHaveBeenCalled();
		expect(handlerMock).toHaveBeenCalledWith(0, 0, 1);

		// test mobile
		const mobileHandlerMock = jest.fn(() => {});

		rerender({
			isFocus: true,
			isMobile: true,
			toggleCandidate: mobileHandlerMock,
		});

		fireEvent.mouseDown(view.container.querySelector(".candidateCell"));
		expect(mobileHandlerMock).not.toHaveBeenCalled();
	});
});
