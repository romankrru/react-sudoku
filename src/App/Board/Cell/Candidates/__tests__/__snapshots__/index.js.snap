// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Candidates /> should show correct candidates 1`] = `
<div>
  <div
    class="candidates"
  >
    <div
      class="candidateCell candidateCellActive"
    >
      <span>
        1
      </span>
    </div>
    <div
      class="candidateCell candidateCellHidden"
    >
      <span>
        2
      </span>
    </div>
    <div
      class="candidateCell candidateCellActive"
    >
      <span>
        3
      </span>
    </div>
    <div
      class="candidateCell candidateCellHidden"
    >
      <span>
        4
      </span>
    </div>
    <div
      class="candidateCell candidateCellHidden"
    >
      <span>
        5
      </span>
    </div>
    <div
      class="candidateCell candidateCellActive"
    >
      <span>
        6
      </span>
    </div>
    <div
      class="candidateCell candidateCellHidden"
    >
      <span>
        7
      </span>
    </div>
    <div
      class="candidateCell candidateCellHidden"
    >
      <span>
        8
      </span>
    </div>
    <div
      class="candidateCell candidateCellHidden"
    >
      <span>
        9
      </span>
    </div>
  </div>
</div>
`;
